Add timezone support

pranavkm opened the pull request:
The way .NET Wasm and .NET Core in *nix read tz information, is by looking at the contents of `/usr/share/zoneinfo/` directory. This directory contains multiple binaries, one for each entry in the tz database: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones. The directory has the same structure as the names of IANA timezones. There's extra content in there, but I verified that the list of regions listed here - https://en.wikipedia.org/wiki/Tz_database#Area is all we need. 

If we distributed the database as-is, Blazor could be configured to download exactly the one file for the current TZ. For instance, there's a file called `/usr/share/zoneinfo/America/Los_Angeles` that corresponds to the `America/Los_Angeles` tz. The tz reported by browsers is IANA string. These files are measured in the few kbs (uncompressed), so it has very little impact on the total download size. 

One downside of downloading a single file is that it's the only TZ available to .NET. Consequently, any attempts to perform TZ conversions other than to UTC would fail (think calendar apps). If we cared about this, we might want to have an additional file that is a single archive of the entire timezone db. 

Another downside is that the TZ db is pretty large (~500 files), so your publish size increases dramatically. We could consider making a build filter that lets you skip tzs that the developer is not interested in. Another option is to produce an archive per region. This means you download extra stuff that isn't used by the app. I tried producing a trivial archive, and some regions like America, and Europe are large (> 20kb) even when compressed.

pranavkm committed:
Timezone draft

SteveSandersonMS commented:
Really great to see this progressing!

Now we know more about the exact compressed size of the data and the number of files, I think we need to consider more broadly the tradeoffs we're making among completeness, usability, and download size.

In particular, now we know the complete brotli-compressed timezone data file (as produced by [@kjpou1's tool](https://github.com/kjpou1/mono-webassembly-zoneinfo)) is just 62kb, I no longer think it's a good tradeoff to split the file at all - not even at the continent level. I think it's a better tradeoff for timezone support to be an all-or-nothing choice. Consider this scenario:

 * @pranavkm is in Redmond and starts using an offline-enabled Blazor PWA - it's a kind of social network for graffiti artists. It caches the `America/Los_Angeles` data file (or maybe the whole of "america" - doesn't matter) and works great, correctly displaying times on an events calendar.
 * Now he boards a plane to Heathrow, ignoring current international travel advice, because he's headed to Bristol on tour to tag some abandoned factories with Banksy
 * In the air, @pranavkm puts his phone into flight mode and sets the device's timezone to the destination so he can start to acclimatize
 * He decides to use the offline-enabled PWA to browse through more of the cached data
 * Since the `Europe\London` timezone data is not available, the app displays nonsense data in the calendar. The app developer had no idea this combination of things could happen, so didn't account for it.
 * Banksy misses out on the opportunity to chill with @pranavkm

More succinctly, it's hard for a developer to realise that *some* timezones may work but not all. Even if they do realise it, it's super hard to discover how to do something about that.

### Suggestion

I think we should use @kjpou1's `zoneinfo.data` in full, and load it by default. This greatly simplifies the developer experience:

 * All timezones work out of the box, like on desktop .NET Core
 * Just one extra file in the deployment bundle
 * Automatically becomes part of the set of files we cache (we include it in `blazor.boot.json` and `service-worker-assets.js`)

I wouldn't be suggesting this if the data file was much bigger, but I think this is a sensible tradeoff given that it compresses down to 62kb.

For the relatively uncommon case where a developer really wants to omit that 62kb and is dedicated enough to discover how to do so, we can offer a MSBuild property like `<BlazorLoadTimezoneData>false</BlazorLoadTimezoneData>` that puts a flag into `blazor.boot.json` that tells the runtime not to load the file. It's fine to still deploy the file either way; it just won't get loaded if you set that flow.

Although splitting the data file (by continent or even individual timezones) would reduce this 62kb to near-zero, the overall payload size difference is slight, and the usability cost of having these gotchas in the product (and our engineering work to make dynamic loading work with PWAs) is significant, so I'd really prefer to avoid splitting if we can.

@danroth27 What do you think? Are you happy with this suggested tradeoff?

### Implementation note

@pranavkm probably doesn't need this info, but since I thought about it, I'm writing it down here anyway: In terms of implementation, the runtime code would need to load `zoneinfo.data` using our `WebAssemblyResourceLoader`. On the MSBuild side, we would want to include that file in `<_BlazorOutputWithTargetPath>` with `BootManifestResourceType` metadata set to `runtime` so that it automatically shows up in the `blazor.boot.json` and `service-worker-assets.js` files.

danroth27 commented:
> What do you think? Are you happy with this suggested tradeoff?

Yup, I think this is the right call.

pranavkm committed:
Add support for reading timezone

pranavkm committed:
Fixup

pranavkm commented:
Still in draft because we need a newer runtime for this, but the changes should be good to go.

pranavkm commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -157,7 +158,7 @@ function addGlobalModuleScriptTagsToDocument(callback: () => void) {
 function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoader, onReady: () => void, onError: (reason?: any) => void) {
   const resources = resourceLoader.bootConfig.resources;
-  const module = {} as typeof Module;
+  const module = (window['Module'] || { }) as typeof Module;
```
This plays slightly better with other scripts that might want to hook up to `Module.preRun`. The initial timezone js file that was shared by the runtime team relied on being able to add a `preRun` task.

pranavkm commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -335,6 +349,52 @@ function attachInteropInvoker(): void {
+  const request = await timeZoneResource.response;
+  let arrayBuffer = await request.arrayBuffer();
+
+  // The contents of the TZ file look like so
```
Here's where we produce the file: https://github.com/dotnet/blazor/pull/2024/files#diff-4a3685061a6b477a3024cb4205ed3a4fR43-R47. Please let me know if I'm missing something important.

SteveSandersonMS replied:
The implementation looks great! Suggestions:

* In the comment block, mention that we are responsible for generating this file, and link to our code in the other repo that does it. Otherwise it looks like it's some sort of industry-standard file format.
* Preferably, factor out this file parsing logic (lines 358-391) into a separate file, since `MonoPlatform.ts` is quite a complicated bundle of semi-related things already

pranavkm commented on src/Components/Web.JS/src/BinaryDecoder.ts:
```diff
@@ -0,0 +1,47 @@
+const uint64HighPartShift = Math.pow(2, 32);
```
The contents of this files were copied from `OutOfProcessRenderBatch.ts`.

pranavkm committed:
Undo changes to standalone

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/blazor build 20200413.2

- Microsoft.AspNetCore.Components.WebAssembly.Runtime: 3.2.0-preview5.20210.3 -> 3.2.0-preview5.20213.2

pranavkm committed:
Regen js scripts

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -178,6 +179,15 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
+  const dotnetTimeZoneResourceName = 'dotnet.timezones.dat';
+
+  let timeZoneResource: LoadingResource | undefined;
+  if (Object.keys(resourceLoader.bootConfig.resources.runtime).indexOf(dotnetTimeZoneResourceName) != -1) {
```
```suggestion
  if (resourceLoader.bootConfig.resources.runtime.hasOwnProperty(dotnetTimeZoneResourceName)) {
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -335,6 +349,52 @@ function attachInteropInvoker(): void {
+  // length of the file. The starting offset of a file can be calculated using the lengths of all files
+  // that appear prior to it.
+
+  const manifestSize = readInt32LE(new Uint8Array(arrayBuffer.slice(0, 4)), 0);
```
Oh, hold on, `ArrayBuffer.slice` actually creates a separate copy of the buffer. So all these calls to `.slice` are very wasteful.

What I think you want instead here is:

```suggestion
  const manifestSize = readInt32LE(new Uint8Array(arrayBuffer, 0, 4));
```

The `Uint8Array` constructor produces a new view over the underlying buffer, but isn't copying anything.

pranavkm replied:
Gotcha. I'd thought this worked like `Span` or `ArraySegment`, strange that a slice produces copies

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -335,6 +349,52 @@ function attachInteropInvoker(): void {
+
+  const manifestSize = readInt32LE(new Uint8Array(arrayBuffer.slice(0, 4)), 0);
+  arrayBuffer = arrayBuffer.slice(4);
+  const manifestContent = new TextDecoder().decode(arrayBuffer.slice(0, manifestSize));
```
I think:

```suggestion
  const manifestContent = new TextDecoder().decode(new Uint8Array(arrayBuffer, 0, manifestSize));
```

... to avoid the unnecessary copy.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -335,6 +349,52 @@ function attachInteropInvoker(): void {
+  arrayBuffer = arrayBuffer.slice(4);
+  const manifestContent = new TextDecoder().decode(arrayBuffer.slice(0, manifestSize));
+  const manifest = JSON.parse(manifestContent) as ManifestEntry[];
+  arrayBuffer = arrayBuffer.slice(manifestSize);
```
```suggestion
  let remainingData = new Uint8Array(arrayBuffer, manifestSize);
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -335,6 +349,52 @@ function attachInteropInvoker(): void {
+
+  const manifestSize = readInt32LE(new Uint8Array(arrayBuffer.slice(0, 4)), 0);
+  arrayBuffer = arrayBuffer.slice(4);
+  const manifestContent = new TextDecoder().decode(arrayBuffer.slice(0, manifestSize));
```
Note that TextDecoder isn't available on Edge older than January 2020: https://caniuse.com/#feat=textencoder

We already have a polyfill in our sources here though - see `src\Rendering\RenderBatch\Utf8Decoder.ts`. This polyfill uses `TextDecoder` when available, and falls back to a JS implementation if not.

It's unfortunate to make `blazor.webassembly.js` bigger by referencing that, but probably is worth it as some customers may still be on Edge that's more than a few months old, and this would be the only incompatible thing for them.

pranavkm replied:
January was the first [public release](https://www.zdnet.com/article/microsofts-chromium-based-edge-browser-to-be-generally-available-january-15-2020) of Edge. My guess is caniuse does not track preview builds of the browser.

SteveSandersonMS replied:
Huh? Edge has existed in pre-Chromium form for many years, and that pre-Chromium version does support WebAssembly.

pranavkm replied:
Oops, TIL

SteveSandersonMS reviewed (approved):
Superb! Thanks for pushing this along and figuring out the whole E2E story for this 🎈 

One final thing: did you verify that the MSBuild stuff here plugs in with @javiercn's precompression as expected? We'd definitely want the timezone data to end up being brotli-compressed.

pranavkm committed:
Changes per PR comments

pranavkm committed:
decodeUtf8

pranavkm commented:
Sizes look about right:

```
-a----         4/14/2020  1:05 AM         469533 dotnet.timezones.dat
-a----         4/14/2020 12:02 PM          63760 dotnet.timezones.dat.br
-a----         4/14/2020 12:02 PM         110398 dotnet.timezones.dat.gz
```

pranavkm committed:
Fixups

pranavkm committed:
Another fixup

pranavkm commented on src/Components/WebAssembly/Server/src/ComponentsWebAssemblyApplicationBuilderExtensions.cs:
```diff
@@ -76,6 +76,7 @@ private static StaticFileOptions CreateStaticFilesOptions(IFileProvider webRootF
             // release builds unless BlazorEnableDebugging is explicitly set to true.
             AddMapping(contentTypeProvider, ".pdb", MediaTypeNames.Application.Octet);
             AddMapping(contentTypeProvider, ".br", MediaTypeNames.Application.Octet);
+            AddMapping(contentTypeProvider, ".dat", MediaTypeNames.Application.Octet);
```
FYI @javiercn. The timezone file that was added is named `dotnet.timezones.dat` which required adding this. I hadn't noticed this issue until I published. Also added the extension to the standalone.web.config.
